File: 873-guess-the-word/guess-the-word.java
/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface Master {
 *     public int guess(String word) {}
 * }
 */
class Solution {
    public void findSecretWord(String[] words, Master master) {
        Random random = new Random();
        int matches = -1;
        List<String> candidates = Arrays.asList(words);
        while (matches != 6) {
            String guess = candidates.get(random.nextInt(candidates.size()));
            matches = master.guess(guess);
            List<String> newCandidates = new ArrayList<>();
            for (String word : candidates) {
                if (matches == getMatches(guess, word)) {
                    newCandidates.add(word);
                }
            }
            candidates = newCandidates;
        }
    }
    private int getMatches(String word1, String word2) {
        int m = 0;
        for (int i = 0, j = 0; i < 6 && j < 6; ++i, ++j){
            if (word1.charAt(i) == word2.charAt(j)){
                m++;
            }
        }
        return m;
    }
}

File: 873-guess-the-word/guess-the-word.py
# """
# This is Master's API interface.
# You should not implement it, or speculate about its implementation
# """
# class Master:
#     def guess(self, word: str) -> int:

class Solution:
    def findSecretWord(self, words: List[str], master: 'Master') -> None:
        words = set(words)
        info = {}
        while len(words) > 0:
            w = words.pop()  # randomly pick a word
            n_match = master.guess(w)
            info[w] = n_match
            # remove all impossible words
            to_remove = set()
            for i in words:
                for j in info:
                    n = 0
                    for k in range(6):
                        if i[k] == j[k]:
                            n += 1
                    if n != info[j]:
                        to_remove.add(i)
                        break
            for i in to_remove:
                if i in words:
                    words.remove(i)